torchmdnet/neighbors/neighbors_cuda_cell.cuh

```diff
@@ -0,0 +1,409 @@
+ * atom in the sorted list
+ */
+static auto sortPositionsByHash(const Tensor& positions, const Tensor& batch,
+                                const Tensor& box_size, const Scalar& cutoff) {
+    const int num_atoms = positions.size(0);
+    Tensor hash_keys = empty({num_atoms}, positions.options().dtype(torch::kInt32));
+    const int threads = 128;
+    const int blocks = (num_atoms + threads - 1) / threads;
+    auto stream = at::cuda::getCurrentCUDAStream();
+    AT_DISPATCH_FLOATING_TYPES(positions.scalar_type(), "assignHash", [&] {
+        scalar_t cutoff_ = cutoff.to<scalar_t>();
+        scalar3<scalar_t> box_size_ = {box_size[0][0].item<scalar_t>(),
+                                       box_size[1][1].item<scalar_t>(),
+                                       box_size[2][2].item<scalar_t>()};
+        assignHash<<<blocks, threads, 0, stream>>>(get_accessor<scalar_t, 2>(positions),
+                                                   get_accessor<int32_t, 1>(hash_keys), box_size_,
+                                                   cutoff_, num_atoms);
+    });
+    // Sort the hash values by the hash keys
+    torch::Tensor sorted_hash_values;
+    torch::Tensor sorted_hash_keys;
+    std::tie(sorted_hash_keys, sorted_hash_values) = torch::sort(hash_keys);
+    Tensor sorted_positions = positions.index_select(0, sorted_hash_values);
+    Tensor sorted_batch = batch.index_select(0, sorted_hash_values);
+    return std::make_tuple(sorted_positions, sorted_batch, sorted_hash_values.to(torch::kInt32));
+}
+
+template <typename scalar_t>
+__global__ void fillCellOffsetsD(const Accessor<scalar_t, 2> sorted_positions,
+                                 const Accessor<int32_t, 1> sorted_indices,
+                                 Accessor<int32_t, 1> cell_start, Accessor<int32_t, 1> cell_end,
+                                 scalar3<scalar_t> box_size, scalar_t cutoff) {
+    // Since positions are sorted by cell, for a given atom, if the previous atom is in a different
+    // cell, then the current atom is the first atom in its cell We use this fact to fill the
+    // cell_start and cell_end arrays
+    const int32_t i_atom = blockIdx.x * blockDim.x + threadIdx.x;
+    if (i_atom >= sorted_positions.size(0))
+        return;
+    const auto pi = fetchPosition(sorted_positions, i_atom);
+    const int3 cell_dim = getCellDimensions(box_size, cutoff);
+    const int icell = getCellIndex(getCell(pi, box_size, cutoff), cell_dim);
+    int im1_cell;
+    if (i_atom > 0) {
+        int im1 = i_atom - 1;
+        const auto pim1 = fetchPosition(sorted_positions, im1);
+        im1_cell = getCellIndex(getCell(pim1, box_size, cutoff), cell_dim);
+    } else {
+        im1_cell = 0;
+    }
+    if (icell != im1_cell || i_atom == 0) {
+        int n_cells = cell_start.size(0);
+        cell_start[icell] = i_atom;
+        if (i_atom > 0) {
+            cell_end[im1_cell] = i_atom;
+        }
+    }
```

Comment:
But you must find the offsets using cell index always. This kernel will work for any hash as long as such hash makes particles in the same cell contiguous, even if two cells have the same hash.
This kernel just looks for places where cell(sortPos[i]) != cell(sortPos[i+1]).

One could also use a Z-order curve for ordering the cell offsets, so that cell_start for a cell is not at cell.x + n.y*(cell.y+n.y*cell.z) but something more sophisticated. It must be slightly faster but I have never explored this.
You only access cell_start once per cell, but potentially access sorted_positions many times per cell. This is why I sort positions but do not bother sorting cell_start.

Reply:
Anyhow using cell index as hash simplifies things, so lets do that. I made it so the hash is assumed to be the cell index.

Reply:
> This kernel will work for any hash as long as such hash makes particles in the same cell contiguous

If two cells have the same hash, the particles won't be contiguous.  The particles from two cells will be interspersed with each other.

Reply:
Yeah you are right, I think we are saying the same thing.